cmd,internal/era: implement `export-history` subcommand

lightclient opened the pull request:
This PR is allows users to export their chain into an an archive format called Era1. It is formulated similarly to the Era[^1] format, which is optimized for reading and distribution CL data. The Era and Era1 format are stricter subsets of a simple [type-length-value](https://en.wikipedia.org/wiki/Type%E2%80%93length%E2%80%93value) scheme called e2store[^2], both developed by the Nimbus team.

The Era format was originally designed to distribute Beacon chain data, but it is clearly desirable to have the same functionality for the EL. A shareable and verifiable archive format is generally considered the first step towards implementing history pruning[^3].

For these reasons, a special flavor of the Era format was developed[^4]: Era1. It's goal is to service EL history *pre-merge*. The reason the format only concerns itself with data before the merge is because post-merge all EL data is encapsulated within the Beacon chain and therefore the existing Era scheme is sufficient.

#### Specification

The format can be summarized with the following expression:

       era1 := Version | block-tuple* | other-entries* | Accumulator | BlockIndex
       block-tuple :=  CompressedHeader | CompressedBody | CompressedReceipts | TotalDifficulty

Each basic element is its own e2store entry:

       Version            = { type: 0x3265, data: nil }
       CompressedHeader   = { type: 0x03,   data: snappyFramed(rlp(header)) }
       CompressedBody     = { type: 0x04,   data: snappyFramed(rlp(body)) }
       CompressedReceipts = { type: 0x05,   data: snappyFramed(rlp(receipts)) }
       TotalDifficulty    = { type: 0x06,   data: uint256(header.total_difficulty) }
       AccumulatorRoot    = { type: 0x07,   data: hash_tree_root(List(HeaderRecord, 8192)) }
       BlockIndex         = { type: 0x3266, data: block-index }

`TotalDifficulty` is little-endian encoded.

`HeaderRecord` is defined in the Portal Network specification[^5].

`BlockIndex` stores relative offsets to each compressed block entry. The format is:

       block-index := starting-number | index | index | index ... | count

All values in the block index are little-endian `uint64`.

`starting-number` is the first block number in the archive. Every index is a defined relative to index's location in the file. The total number of block entries in the file is recorded in count.

Due to the accumulator size limit of 8192, the maximum number of blocks in an Era batch is also 8192. This is also the value of `SLOTS_PER_HISTORICAL_ROOT`[^6] on the Beacon chain, so it is nice to align on the value.

#### Verification

There are two verification paths.

##### Accumulator Roots

The accumulator roots are hash tree roots of `HeaderRecord`s[^5]. This is a relatively format-agnostic verification method. So if Era1 changes in the future, the accumulator values will not as it is shared with the Portal Network. However, verifying the accumulators is expensive because the underlying transactions and receipts need to be verified against their headers before the overall accumulator is checked. Once this is complete, the `HeaderRecord`s can be generated `htr`'d. The `era` binary is provided by `cmd/era`. For a full list of accumulator values, see [`accumulators.md`](https://gist.github.com/kdeme/b4b3c572e8413a35881b4b08e72181f2).

```console
$ era verify accumulators.txt
```


##### Checksums

*Note: checksums are platform dependent due to the lack of byte-for-byte conformance of snappy. They are for convenience if clients wish to distribute them.*

In order to provide fast verification of integrity, we also provide checksum values of each Era1 file. The checksum is simply the `sha256` digest of the Era1 file. This requires trusting a third party's checksum values.

```console
$ sha256sum era/mainnet-00000-5ec1ffb8.era1
407c4c0eecd727ba70368bab3750043b722e00abb62c1bb44d4557d085fe2173
```

See [`checksums.txt`](https://gist.github.com/lightclient/183e920728e1671800411efb134dbbc0) for a full list of checksum values.

[^1]: https://github.com/status-im/nimbus-eth2/blob/stable/docs/e2store.md#era-files
[^2]: https://github.com/status-im/nimbus-eth2/blob/stable/docs/e2store.md
[^3]: https://eips.ethereum.org/EIPS/eip-4444
[^4]: https://hackmd.io/@arnetheduck/H15vMzx2s
[^5]: https://github.com/ethereum/portal-network-specs/blob/master/history-network.md#the-header-accumulator
[^6]: https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#time-parameters

holiman commented:
Is this PR superseding #25325 ?

holiman commented on cmd/utils/cmd.go:
```diff
@@ -287,6 +290,54 @@ func ExportAppendChain(blockchain *core.BlockChain, fn string, first uint64, las
+		if err != nil {
+			return err
+		}
+		defer fh.Close()
```
`defer` inside a loop is usually not great, they will pile up and you will close `N` files when the whole thing is done and the method ends.

lightclient replied:
Good point. Updated.

holiman commented on cmd/utils/cmd.go:
```diff
@@ -287,6 +290,54 @@ func ExportAppendChain(blockchain *core.BlockChain, fn string, first uint64, las
+			if receipts == nil {
+				return fmt.Errorf("export failed on #%d: receipts not found", nr)
+			}
+			if err := w.Add(block, receipts); err != nil {
```
what about withdrawals?

lightclient replied:
This is specifically for pre-merge data, not sure if it makes sense to also support withdrawals? I guess it doesn't really hurt.

holiman commented on internal/e2store/e2store.go:
```diff
@@ -0,0 +1,122 @@
+	length += uint64(b[6]) << 32
+	length += uint64(b[7]) << 40
+
+	val := make([]byte, length)
```
This is a bit unsafe. length can be `255 PB`, I think this should be capped to maybe 2GB or something.

holiman replied:
Tangent: why so large length even?

lightclient replied:
Yeah I realized this, was trying to understand how `rlp` handles this since the format technically supports strings of size `2^((0xc0-0xb7)*8) `. Seems like it tries to interrogate the underlying type to see if it can figure out how much data is available to be read. 

Is there a better way to deal with this? Suppose `length` for whatever reason is extremely large (large than we can put in memory)? I kind of think we should not support that case, but it would be better fail gracefully than panic here.

lightclient commented:
Yes this supersedes #25325!

holiman commented on cmd/utils/cmd.go:
```diff
@@ -332,8 +335,13 @@ func ExportHistory(bc *core.BlockChain, dir string, first, last, step uint64) er
+			log.Info("Exporting blocks", "exported", i, "elapsed", common.PrettyDuration(time.Since(start)))
+			reported = time.Now()
+		}
+		fh.Close()
```
Not sufficient, you have other `return`s. So either you can salt with `fh.Close()` at all return locations, you you can turn the whole clause within the outer `for` loop into a function -- and then you can use `defer` again.

lightclient replied:
ah yes duh - okay I put it into a function now :+1:

holiman commented on cmd/utils/cmd.go:
```diff
@@ -332,8 +335,13 @@ func ExportHistory(bc *core.BlockChain, dir string, first, last, step uint64) er
 			return fmt.Errorf("export failed to finalize %s: %w", fn, err)
 		}
-		log.Info("Exported blockchain to", "dir", dir, "fn", fn)
+		if time.Since(reported) >= 10*time.Second {
```
`8s` is the go-ethereum-idiomatic value to use in these situations :trollface:

MariusVanDerWijden commented on cmd/geth/chaincmd.go:
```diff
@@ -362,6 +373,39 @@ func exportChain(ctx *cli.Context) error {
+// exportHistory exports chain history in Era archives at a specified
+// directory.
+func exportHistory(ctx *cli.Context) error {
+	if ctx.Args().Len() != 3 {
```
This looks wrong to me, your checking if args.len == 3, but access 4 elements ("<dir> <first> <last> <step>)

lightclient replied:
Ah right, I added step in later.

arnetheduck commented:
> `    Version            = { type: 0x3265, data: nil }`

careful with the byte order here - types are arrays of 2 bytes (not integers) to avoid endian-ness issues - in the e2store spec, the version is `type: [0x65, 0x32]` for this reason

>    Accumulator        = { type: 0x05,   data: hash_tree_root(blockHashes, 8192) }

I wonder if we should store the full tree here or add the list of hashes as a separate entry - in the beacon chain, the root of each block is available from the state allowing the verifier to check blocks one by one - when a file is trusted, this also acts as a quick way to construct a hash->number index, without having to _actually_ hash the data.

lightclient commented:
> I wonder if we should store the full tree here or add the list of hashes as a separate entry 

You mean store the full list of header records (I updated the defn btw, I realize it should have been a [`List(HeaderRecord, 8192)`](https://github.com/ethereum/portal-network-specs/blob/master/history-network.md#the-header-accumulator) from the portal network)? I think I prefer the list of hashes as a separate entry over the full tree, but maybe I don't fully understand.

lightclient commented:
If you want to run this yourself and test it, try the following:

```console
$ geth export-history eras 0 15537393 8192
```

This will export all the pre-merge blocks. For me, it takes around 4.5 hrs. You can verify them by running:

```console
$ wget https://gist.githubusercontent.com/lightclient/528b95ffe434ac7dcbca57bff6dd5bd1/raw/fd660cfedb65cd8f133b510c442287dc8a71660f/roots.txt
$ era verify roots.txt
```

This takes around 3.5 hours.

arnetheduck commented:
>  I think I prefer the list of hashes as a separate entry over the full tree, but maybe I don't fully understand.

excellent - either list of full tree works well, full list is certainly enough!

lightclient commented:
@arnetheduck I implemented a the hashes section in a [separate branch](https://github.com/lightclient/go-ethereum/commit/869dc227507d7c401706129d14079ec9e38813e8). Not sure if it makes sense to include here right now, can always add later (backwards compatible) if it's something that we really need.

--

I think this PR is ready to be reviewed again. @holiman, I took a look at changing this PR to only use the freezer, but ultimately I need the genesis block too (which IIUC is not available from freezer), so would need to already open the datadir to get that. I also am getting the network id to determine the naming scheme (would be easy to replace this with cli parameter though). Here is the [branch](https://github.com/lightclient/go-ethereum/commit/3ae5a6c9f4f5a07de58b836265affa59b6456d38) associated with the attempt. Let me know if you still feel strongly about switching to freezer-only format and maybe we can work out a different way!

lightclient commented:
Per triage: going to break out the Era logic into a separate repo.

lightclient commented:
Okay I've moved all the Era logic to https://github.com/lightclient/era! This PR should be good to go now.

MariusVanDerWijden commented on cmd/utils/cmd.go:
```diff
@@ -287,6 +291,91 @@ func ExportAppendChain(blockchain *core.BlockChain, fn string, first uint64, las
+	for i := uint64(first); i < last; i += step {
+		gen := func() error {
+			// Open file for Era.
+			fn := path.Join(dir, fmt.Sprintf("%s-%05d.era", network, i/step))
```
The nimbus spec proposes the following naming scheme: `<config-name>-<era-number>-<era-count>-<short-historical-root>.era` afaict https://github.com/status-im/nimbus-eth2/blob/stable/docs/e2store.md

lightclient replied:
Ah good point - I've fixed this in the latest. The example shown doesn't have `<era-count>` for some reason. I recall when I download some beacon era files they also didn't so something may be out of sync there. I've omitted it now for that reason.

MariusVanDerWijden commented on cmd/geth/chaincmd.go:
```diff
@@ -117,6 +118,17 @@ Optional second and third arguments control the first and
+		Action:    exportHistory,
+		Name:      "export-history",
+		Usage:     "Export blockchain history to Era archives",
+		ArgsUsage: "<dir> <first> <last> <step>",
```
```suggestion
		ArgsUsage: "<dir> <first> <last>",
```

rjl493456442 commented on cmd/utils/cmd.go:
```diff
@@ -287,6 +291,92 @@ func ExportAppendChain(blockchain *core.BlockChain, fn string, first uint64, las
 	return nil
 }
 
+func ExportHistory(bc *core.BlockChain, dir string, first, last, step uint64) error {
```
It's good to have a separate function operating on the blockchain object directly. 
We might be enable to provide a RPC function to dump the chain history without
shutdown.

lightclient replied:
Sorry looks like I missed this! Are you referring to creating a function like `ExportN`? If so, two thoughts:

* this is a very special use format for pre-merge data, it isn't something that really needs to be computed by many people so I don't think it is useful to increase the surface of the blockchain api with additional methods
* the era1 format in general should probably be exposed as little as possible due to extremely limited usefulness

holiman commented:
Here's an experiment

```diff
diff --git a/cmd/utils/cmd.go b/cmd/utils/cmd.go
index 15875602d8..2ab4271f00 100644
--- a/cmd/utils/cmd.go
+++ b/cmd/utils/cmd.go
@@ -46,6 +46,7 @@ import (
 	"github.com/ethereum/go-ethereum/rlp"
 	"github.com/lightclient/era"
 	"github.com/urfave/cli/v2"
+	"math/rand"
 )
 
 const (
@@ -291,6 +292,16 @@ func ExportAppendChain(blockchain *core.BlockChain, fn string, first uint64, las
 	return nil
 }
 
+func flipBit(data []byte) int {
+	idx := rand.Intn(len(data) * 8)
+	if data[idx/8] == 0x13 {
+		fmt.Printf("magic\n")
+	}
+	fmt.Printf("flipping: data[%d]: %x (block size %d)\n", idx/8, data[idx/8], len(data))
+	data[idx/8] ^= 1 << (idx % 8)
+	return idx
+}
+
 func ExportHistory(bc *core.BlockChain, dir string, first, last, step uint64) error {
 	log.Info("Exporting blockchain history", "dir", dir)
 	if head := bc.CurrentBlock().Number.Uint64(); head < last {
@@ -343,6 +354,7 @@ func ExportHistory(bc *core.BlockChain, dir string, first, last, step uint64) er
 				if err != nil {
 					return fmt.Errorf("encode block failed #%d: %w", nr, err)
 				}
+
 				r := bc.GetReceiptsByHash(b.Hash())
 				if r == nil {
 					return fmt.Errorf("export failed on #%d: receipts not found", nr)
@@ -355,6 +367,11 @@ func ExportHistory(bc *core.BlockChain, dir string, first, last, step uint64) er
 					hash = b.Hash().Bytes()[:]
 					td   = bc.GetTd(b.Hash(), nr)
 				)
+				if nr == 2 {
+					idx := flipBit(block)
+					fmt.Printf("flipped: data[%d]: %x block %d txs: %d\n", idx/8, block[idx/8], b.NumberU64(),
+						len(b.Body().Transactions))
+				}
 				if err := w.Add(block, receipts, hash, td); err != nil {
 					return err
 				}
diff --git a/cmd/utils/history_test.go b/cmd/utils/history_test.go
index 61961a0b1b..6cf8c5d06e 100644
--- a/cmd/utils/history_test.go
+++ b/cmd/utils/history_test.go
@@ -40,6 +40,14 @@ var (
 	step  uint64 = 16
 )
 
+func fill(size int) []byte {
+	d := make([]byte, size)
+	for i := range d {
+		d[i] = 0x13
+	}
+	return d
+}
+
 func TestHistoryExporter(t *testing.T) {
 	var (
 		key, _  = crypto.HexToECDSA("b71c71a67e1177ad4e901695e1b4b9ee17ae16c6668d313eac2f96dbcda3f291")
@@ -64,7 +72,7 @@ func TestHistoryExporter(t *testing.T) {
 			Gas:        50000,
 			To:         &common.Address{0xaa},
 			Value:      big.NewInt(int64(i)),
-			Data:       nil,
+			Data:       fill(100),
 			AccessList: nil,
 		})
 		if err != nil {
```
What this does, is it modifies the data being saved. First, I attempted to just flip a bit on disk, but snappy detects that. This instead flips a bit in the rlp-representation of a `block`. You may also notice that I added data on the transactions, because I suspected that tx-data modifications would be unverified (I was right:) ) 

If you run this several times, e.g. 
```
[user@work utils]$ go test . -run History --count 10 > foo
```
Most times, it will detect the error: 
```
INFO [03-24|08:44:41.258] Exporting blockchain history             dir=/tmp/history-export-test35840071
flipping: data[403]: 0 (block size 723)
flipped: data[403]: 2 block 2 txs: 1
INFO [03-24|08:44:41.270] Exported blockchain to                   dir=/tmp/history-export-test35840071
--- FAIL: TestHistoryExporter (0.18s)
    history_test.go:131: block hash mistmatch 2: want 0x888efbdbd17c2ae9ed411871df2832936dbec0e998da10a9cc0cc2429f2770a9, got 0x1982f8681594de480fdc32de7d7ae3f97a2c25b0bd9277e6a54c3ee7957673b7
```
But you will find a couple of these 
```
INFO [03-24|08:44:41.047] Exporting blockchain history             dir=/tmp/history-export-test2902024754
magic
flipping: data[644]: 13 (block size 723)
flipped: data[644]: 53 block 2 txs: 1
INFO [03-24|08:44:41.066] Exported blockchain to                   dir=/tmp/history-export-test2902024754

```
Where the modification was not detected.

--- 
In my opinion, this is not just a case of "Oh, @lightclient missed a detail in the implementation.". This type of flaw is a symptom of a larger underlying issue: 
- the `era` -format not a standalone dataformat with integrity guarantees
- The `era`-format is too heavily tied to the application data (blockchain types), it is very unclear what guarantees are offered by the era-format validations, and which checks need to be made by the parent application. 

As I said, trying to bitflip on disk was immediately detected by snappy. Because `snappy`, by comparison, is fully standalone, and it does not matter what I store on it. The guarantees given by the snappy-reader are not vague at all. 

I think we're going to shoot ourselves in the foot with `era`. I'm very tempted to do this exact same excercise with the existing implementations on the CL side.

arnetheduck commented:
> As I said, trying to bitflip on disk was immediately detected by snappy. Because snappy, by comparison, is fully standalone, and it does not matter what I store on it. The guarantees given by the snappy-reader are not vague at all.

Per https://github.com/google/snappy/blob/main/format_description.txt, snappy itself contains no data protection features (no crc, no hashes etc). 

If instead you are referring to the snappy framing format, [some](https://github.com/google/snappy/blob/main/framing_format.txt#L39), but notably not all, of the data is protected by CRC-32 - this is analogous to era files: some of the data is protected by a type-defined hash and verification thereof is part of the specification (https://github.com/status-im/nimbus-eth2/blob/stable/docs/e2store.md#verifying-era-files) - verifying implementations can plausibly detect everything from bit-flips to malicious modifications (unlike snappy where malicious actors trivially can make the crc32 pass).

In addition to the hash-based chain verification, payloads in era files are actually snappy-framing-compressed meaning that they inherit the crc-32 protection from snappy itself (analogous to snappy itself where de-facto only payloads are covered by crc-32) - this means era files practically offer the same bit-flip protection as snappy itself.

There does exist one notable difference: to fully verify an era file using the heavier hash-verification, you need an out-of-band reference (ie knowledge of the hash of the accumulator).

holiman commented:
Yep, I grant that I attributed too much to snappy, but the point of snappy being data-agnostic still stands. 

---

Verifying era-files https://github.com/status-im/nimbus-eth2/blob/stable/docs/e2store.md#verifying-era-files

> the state is loadable and consistent with the given runtime configuration

What does that mean? We have no state in the exported blockchain files, do we? And if we did, it sounds like "verify the state". Like, "verify an era-file" == "verify the sequence of blocks" == "verify the blockchain" == "import the blocks" ?

The specification there, to me, just highlights the same problem. The `era`-format is heavily tied to application logic, and you wrote it for CL data. Which means that the `era` format is not canonically defined for EL datatypes, and there's no single canonical specification of how to verify an (EL) era-file.

arnetheduck commented:
> Yep, I grant that I attributed too much to snappy, but the point of snappy being data-agnostic still stands.

Indeed - this is a small benefit of snappy, though I believe the store inherits it for the majority of payloads (indeed because we use snappy throughout).

> and you wrote it for CL data.

Ah, this circles back to a thing we were discussing with @lightclient - ie we're aiming to produce an era-file-equivalent spec for pre-merge eth1 data - the outcome is _not_ an era file but rather an .. `era1` file maybe? need a good name here.

However, the main idea was to use the `e2store` "framing" so as to recycle the basics of the format that "underpin" era files (compress things with snappy so we don't introduce additional dependencies compared to a "standard" ethereum client today, use TLV to make it extensible but easy to consume, allow `O(1)` lookups by block number) - however, the files we're exporting here are not era files per se - think of "era" as the application-specific instantiation of e2store.

That said, it would be possible to make `e2store` self-verifiable at the base layer by simply adding a record type which does something like "crc32 of the data from the last checksum", if this is a desired "base-layer" application-agnostic feature - the question hasn't come up though simply because of the basic self-verification offered by snappy already as well as the deeper application-specific secure-hash-driven verification in the application layer.

Would this solve the concern?

arnetheduck commented:
> What does that mean? 

I hope the above comment clears it up: ie "era" is the application-specific instantiation so we can rely on the CL state.

In the pre-merge era1 format, the `Accumulator` serves the same purpose, ie that's the piece of information you need to do "application-level" verification.

lightclient commented:
Okay updated this PR to fix a handful of things:
* there is now a `checksums.txt` file generated, which can be checked against published checksum values for fast data integrity verification
* filenames follow era spec
* export takes around 6 hr and import around 9.5 hr

I think it's ready again for a wave of review!

MariusVanDerWijden commented on internal/era/verify.go:
```diff
@@ -0,0 +1,64 @@
+package era
+
+// func (r *Reader) Verify() error {
```
Looks like this code was moved to era/main and is not needed anymore

lightclient replied:
done

MariusVanDerWijden commented on cmd/e2store/main.go:
```diff
@@ -0,0 +1,117 @@
+// open opens the e2store at the provided path.
+func open(ctx *cli.Context, argCount int) (*os.File, error) {
+	if ctx.Args().Len() != argCount {
+		return nil, fmt.Errorf("missing file argument")
```
You are checking here the length of args, but the error doesn't really match.
For example if you call find with only one argument or three argument, the error thats returned is missing file argument. Better imo would be something like "invalid number of arguments, got %v want %v"

lightclient replied:
deleted the whole file 🤝

MariusVanDerWijden commented on cmd/era/main.go:
```diff
@@ -0,0 +1,334 @@
+var (
+	dirFlag = &cli.StringFlag{
+		Name:  "dir",
+		Usage: "directory storing all relevant Era files",
```
inconsistent capitalization, sometimes you use Era sometimes era

lightclient replied:
fixed

MariusVanDerWijden commented on cmd/era/main.go:
```diff
@@ -0,0 +1,334 @@
+func info(ctx *cli.Context) error {
+	epoch, err := strconv.ParseUint(ctx.Args().First(), 10, 64)
+	if err != nil {
+		return fmt.Errorf("invalid block number: %w", err)
```
invalid epoch

MariusVanDerWijden commented on cmd/era/main.go:
```diff
@@ -0,0 +1,334 @@
+	if epoch >= uint64(len(entries)) {
+		return nil, fmt.Errorf("epoch out-of-bounds: have %d, want %d", len(entries), epoch)
+	}
+	return os.Open(path.Join(dir, entries[epoch]))
```
This only works if the era files are continuous right? I think if I have an `aaa.era`, or I'm missing a file it would open the wrong file

lightclient replied:
Yeah `era.ReadDir` only returns non-error if the eras are sequential... not ideal, but I think solving this now is unnecessarily complicated when it isn't currently useful to have gaps in eras.

MariusVanDerWijden replied:
You could return a map instead of an array with the filenames, but kinda agree that its unnecessary if we assume gaps in eras are impossible

MariusVanDerWijden commented on cmd/era/main.go:
```diff
@@ -0,0 +1,334 @@
+	//   * the block index is constructed correctly
+	//   * the starting total difficulty value is correct
+	//   * the accumulator is correct by recomputing it locally,
+	//     which verfies the blocks are all correct (via hash)
```
```suggestion
	//     which verifies the blocks are all correct (via hash)
```

lightclient replied:
fixed

MariusVanDerWijden commented on cmd/geth/chaincmd.go:
```diff
@@ -335,26 +361,108 @@ func exportChain(ctx *cli.Context) error {
+			if err != nil {
+				return fmt.Errorf("error reading %s: %w", dir, err)
+			}
+			if 0 < len(entries) {
```
Could you turn this around?

lightclient replied:
fixed

MariusVanDerWijden commented on cmd/geth/main.go:
```diff
@@ -198,7 +198,9 @@ func init() {
 		// See chaincmd.go:
 		initCommand,
 		importCommand,
+		importHistoryCommand,
```
I would switch those, make the import/export pairs next to each other

MariusVanDerWijden commented on cmd/utils/cmd.go:
```diff
@@ -224,6 +229,88 @@ func ImportChain(chain *core.BlockChain, fn string) error {
+		imported = 0
+		forker   = core.NewForkChoice(chain, nil)
+	)
+	for i, filename := range entries {
```
Iterate over the checksums, since len(checksums) is guaranteed to be leq entries

lightclient replied:
I just made the check ensure they are equal - it seems cleaner and then you don't have a confusing situation where you though you would import all the era files, but you didn't because you missing a checksum value

MariusVanDerWijden commented on cmd/utils/cmd.go:
```diff
@@ -224,6 +229,88 @@ func ImportChain(chain *core.BlockChain, fn string) error {
+
+	// Manually update the head snap block so that when user restarts geth,
+	// it is able to begin syncing from the last imported block.
+	rawdb.WriteHeadFastBlockHash(db, chain.CurrentHeader().Hash())
```
Shouldn't this be written in blockchain/InsertReceiptChain already? (I might be wrong)

lightclient replied:
Yeah you're right, it is now, at some point I was doing something a bit different and had to write in manually.

MariusVanDerWijden commented on cmd/utils/cmd.go:
```diff
@@ -293,6 +380,82 @@ func ExportAppendChain(blockchain *core.BlockChain, fn string, first uint64, las
+				w   = era.NewBuilder(buf)
+			)
+			for j := uint64(0); j < step && j <= last-i; j++ {
+				nr := i + j
```
```suggestion
				number := i + j
```

lightclient replied:
I think `n` or `nr` is okay here, `number` is rather verbose

MariusVanDerWijden commented on cmd/utils/cmd.go:
```diff
@@ -293,6 +380,82 @@ func ExportAppendChain(blockchain *core.BlockChain, fn string, first uint64, las
+			}
+			return nil
+		}
+		if err := gen(); err != nil {
```
Whats the advantage of making gen a function? Is it to parallelize it in the future?

MariusVanDerWijden replied:
I would pull the 8 second logging out into the loop, maybe even get rid of the fun

lightclient replied:
ahh yeah i used to have it so a bunch of `defer`s wouldn't pile up, but those are gone now so the whole function can go

MariusVanDerWijden commented on cmd/utils/history_test.go:
```diff
@@ -0,0 +1,200 @@
+				t.Fatalf("blocks out of order: want %d, got %d", want, got)
+			}
+			if want.Hash() != b.Hash() {
+				t.Fatalf("block hash mistmatch %d: want %s, got %s", i+j, want.Hash().Hex(), b.Hash().Hex())
```
```suggestion
				t.Fatalf("block hash mismatch %d: want %s, got %s", i+j, want.Hash().Hex(), b.Hash().Hex())
```

MariusVanDerWijden commented on cmd/utils/history_test.go:
```diff
@@ -0,0 +1,200 @@
+	}
+
+	// Read checksums.
+	b, err := os.ReadFile(path.Join(dir, "checksums.txt"))
```
The variable b is  used for different things 3 times in this function, twice for files, once for a block

MariusVanDerWijden commented on cmd/utils/history_test.go:
```diff
@@ -0,0 +1,200 @@
+	}
+}
+
+func BenchmarkHistoryImport(b *testing.B) {
```
I get the intention of this benchmark, benchmark the import of a certain export you have locally. I think benchmarks should be self contained, so you don't need some extra piece of data.

MariusVanDerWijden replied:
WDYT about something like this:

```

func BenchmarkImportAndExport(b *testing.B) {
	var (
		key, _  = crypto.HexToECDSA("b71c71a67e1177ad4e901695e1b4b9ee17ae16c6668d313eac2f96dbcda3f291")
		address = crypto.PubkeyToAddress(key.PublicKey)
		genesis = &core.Genesis{
			Config: params.TestChainConfig,
			Alloc:  core.GenesisAlloc{address: {Balance: big.NewInt(1000000000000000000)}},
		}
		signer = types.LatestSigner(genesis.Config)
	)

	// Generate chain.
	db, blocks, _ := core.GenerateChainWithGenesis(genesis, ethash.NewFaker(), int(b.N), func(i int, g *core.BlockGen) {
		if i == 0 {
			return
		}
		tx, err := types.SignNewTx(key, signer, &types.DynamicFeeTx{
			ChainID:    genesis.Config.ChainID,
			Nonce:      uint64(i - 1),
			GasTipCap:  common.Big0,
			GasFeeCap:  g.PrevBlock(0).BaseFee(),
			Gas:        50000,
			To:         &common.Address{0xaa},
			Value:      big.NewInt(int64(i)),
			Data:       nil,
			AccessList: nil,
		})
		if err != nil {
			b.Fatalf("error creating tx: %v", err)
		}
		g.AddTx(tx)
	})

	// Initialize BlockChain.
	chain, err := core.NewBlockChain(db, nil, genesis, nil, ethash.NewFaker(), vm.Config{}, nil, nil)
	if err != nil {
		b.Fatalf("unable to initialize chain: %v", err)
	}
	if _, err := chain.InsertChain(blocks); err != nil {
		b.Fatalf("error insterting chain: %v", err)
	}

	// Make temp directory for era files.
	dir, err := os.MkdirTemp("", "history-export-test")
	if err != nil {
		b.Fatalf("error creating temp test directory: %v", err)
	}
	defer os.RemoveAll(dir)

	b.ResetTimer()
	// Export history to temp directory.
	if err := ExportHistory(chain, dir, 0, uint64(b.N), step); err != nil {
		b.Fatalf("error exporting history: %v", err)
	}
	b.StopTimer()

	// Now import Era.
	freezer := b.TempDir()
	db2, err := rawdb.NewDatabaseWithFreezer(rawdb.NewMemoryDatabase(), freezer, "", false)
	if err != nil {
		panic(err)
	}
	b.Cleanup(func() {
		db2.Close()
	})

	genesis.MustCommit(db2)
	imported, err := core.NewBlockChain(db2, nil, genesis, nil, ethash.NewFaker(), vm.Config{}, nil, nil)
	if err != nil {
		b.Fatalf("unable to initialize chain: %v", err)
	}
	b.StartTimer()
	if err := ImportHistory(imported, db2, dir, "mainnet"); err != nil {
		b.Fatalf("failed to import chain: %v", err)
	}
	if have, want := imported.CurrentHeader(), chain.CurrentHeader(); have.Hash() != want.Hash() {
		b.Fatalf("imported chain does not match expected, have (%d, %s) want (%d, %s)", have.Number, have.Hash(), want.Number, want.Hash())
	}
}
```

lightclient replied:
I think I can just delete this - I had a benchmark similar to this before, but changed it to run the bench on some actual mainnet blocks to get a more realistic bench.

MariusVanDerWijden commented on cmd/utils/cmd.go:
```diff
@@ -224,6 +229,88 @@ func ImportChain(chain *core.BlockChain, fn string) error {
+	if err != nil {
+		return fmt.Errorf("unable to read checksums.txt: %w", err)
+	}
+	if len(checksums) <= len(entries) {
```
```suggestion
	if len(checksums) > len(entries) {
```

MariusVanDerWijden commented on go.mod:
```diff
@@ -100,10 +101,12 @@ require (
 	github.com/kr/text v0.2.0 // indirect
 	github.com/mattn/go-runewidth v0.0.9 // indirect
 	github.com/matttproud/golang_protobuf_extensions v1.0.4 // indirect
+	github.com/minio/sha256-simd v1.0.0 // indirect
```
This is not used anymore I think (and shouldn't)

lightclient replied:
This is from `fastssz`

MariusVanDerWijden commented on internal/era/e2store/e2store.go:
```diff
@@ -0,0 +1,145 @@
+	"io"
+)
+
+var headerSize = 8
```
```suggestion
// Every e2 header is 8 bytes long.
var headerSize = 8
```

MariusVanDerWijden commented on internal/e2store/e2store.go:
```diff
@@ -0,0 +1,150 @@
+}
+
+// CurrentOffset returns the current offset of the underlying writer.
+func (w *Writer) CurrentOffset() (int64, error) {
```
This seems unused

lightclient replied:
you're right

MariusVanDerWijden commented on internal/era/e2store/e2store.go:
```diff
@@ -0,0 +1,145 @@
+// An entry is encoded in a type-length-value format. The first 8 bytes of the
+// record store the type (2 bytes), the length (4 bytes), and some reserved
+// data (2 bytes). The remaining bytes store b.
+func (w *Writer) Write(typ uint16, b []byte) (int, error) {
```
Isn't most of this a copy-paste from `internal/e2store/e2store.go`

MariusVanDerWijden replied:
I would vote for this to stay, and the internal/e2store to go

lightclient replied:
yep, look like it didn't get properly deleted when moving some things around

lightclient commented on cmd/utils/cmd.go:
```diff
@@ -224,6 +229,88 @@ func ImportChain(chain *core.BlockChain, fn string) error {
+		r := era.NewReader(bytes.NewReader(b))
+		for j := 0; ; j += 1 {
+			n := i*era.MaxEra1Size + j
+			b, r, err := r.Read()
```
fixed

MariusVanDerWijden commented on internal/era/era.go:
```diff
@@ -0,0 +1,553 @@
+
+// NewReader returns a new Reader instance.
+func NewReader(r io.ReadSeeker) *Reader {
+	// qq: does it make more sense to read the metadata here and
```
Hmm usually I would leave it this way, but since you need to have two additional checks because of it, I kinda lean to removing it

lightclient replied:
removed!

fjl commented on internal/era/builder.go:
```diff
@@ -0,0 +1,227 @@
+	b.tds = append(b.tds, td)
+
+	// Small helper to take care snappy encoding and writing e2store entry.
+	snappyWrite := func(typ uint16, in []byte) error {
```
This could be a regular method.

fjl commented on internal/era/era.go:
```diff
@@ -0,0 +1,370 @@
+		if err != nil {
+			return off, err
+		}
+		off += e2store.HeaderSize + int64(length)
```
It's not great that knowledge of the format escapes to another package here. I think it'd be better to have 
a function in Reader that returns the next offset.

lightclient replied:
This is a good point. I've fixed it by creating a method `LengthAt` which returns the overall length of the entry (including header) at the given offset.

holiman commented on internal/era/era.go:
```diff
@@ -0,0 +1,370 @@
+	if e.Type != TypeCompressedHeader {
+		return nil, fmt.Errorf("expected header entry, got %x", e.Type)
+	}
+	return io.ReadAll(snappy.NewReader(bytes.NewReader(e.Value)))
```
Since you will read the entire thing via io.ReadAll, why not just use 
```suggestion
	return snappy.Decode(nil, e.Value)
```

holiman replied:
Ah, it seems that you are encodign this using the snappy stream format, not the snappy block format. Is that a deliberate decision?

holiman replied:
Looking at how you implemented `snappyWrite`, it seems a bit like you are using the stream-format but manipulating it into block-format, using `Reset`  to restart new streams every time you might otherwise have used a block: 
```
	var (
		buf = b.buf
		s   = b.snappy
	)
	buf.Reset()
	s.Reset(buf)
```

holiman replied:
TLDR; I think this would be simpler if you just use `snappy.Encode` and `snappy.Decode`

lightclient replied:
Yes exactly. So the [regular Era format](https://github.com/status-im/nimbus-eth2/blob/613f4a9a50c9c4bd8568844eaffb3ac15d067e56/docs/e2store.md#why-snappy-sz-framed-compression) is specified to use snappy stream encoding. On the beacon chain this makes sense, because it is the [same format ](https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/p2p-interface.md#ssz-snappy-encoding-strategy)they use for p2p.

I naively followed suit. Now that you bring it up though, it doesn't seem there is much benefit to use snappy stream for us? Our p2p is snappy block. At the same time, Era1 isn't exactly optimized for p2p requests. Post-merge Era files are also snappy stream encoded anyways, so it doesn't save us too much.

arnetheduck replied:
>  If you receive 100 blocks that are frame (stream) -compressed,

Each block in the response is individually stream-compressed which among other things allows capping the amount of memory needed to decompress them. Ditto era files: all compression is done per "entry" so that we can retain O(1) access to all data. Framed encoding puts a cap on how much memory is needed for compression/decompression - we made a mistake on the CL side using block encoding at all (it's used in gossip, unfortunately).

Frame encoding also comes with a mime type and CRC32 which allows detecting simple corruption trivially - I'd strongly prefer that the eth1 format follows suite here and uses frame encoding, if not only for the simple reason that it makes life easier to have the same guarantees in both formats.

holiman commented on internal/era/era.go:
```diff
@@ -0,0 +1,370 @@
+	if e.Type != TypeCompressedBody {
+		return nil, fmt.Errorf("expected body entry, got %x", e.Type)
+	}
+	return io.ReadAll(snappy.NewReader(bytes.NewReader(e.Value)))
```
```suggestion
	return return snappy.Decode(nil, e.Value)
```

holiman commented on internal/era/era.go:
```diff
@@ -0,0 +1,343 @@
+}
+
+// ReadBodyRLP reads the block body number n RLP.
+func (r *Reader) ReadBodyRLP(n uint64) ([]byte, error) {
```
The `ReadXXRLP` methods, they're not actually required externally..? If so, better make them unexported for now IMO

holiman commented on internal/era/era.go:
```diff
@@ -0,0 +1,343 @@
+}
+
+// Read reads one (block, receipts) tuple from an Era1 archive.
+func (r *Reader) Read() (*types.Block, types.Receipts, error) {
```
I think this method should be called `Next()`, since it's kind of like an iterator, has an internal state which progresses every time you call it

lightclient replied:
done in 3551a7c8f7ca08c0bac18bf618434388dd205500

holiman commented:
Here's a interesting experiment: 
```diff 
diff --git a/cmd/utils/history_test.go b/cmd/utils/history_test.go
index e24a68bbd6..1e0f5555c6 100644
--- a/cmd/utils/history_test.go
+++ b/cmd/utils/history_test.go
@@ -19,6 +19,7 @@ package utils
 import (
 	"bytes"
 	"crypto/sha256"
+	"fmt"
 	"io"
 	"math/big"
 	"os"
@@ -114,12 +115,16 @@ func TestHistoryImportAndExport(t *testing.T) {
 		if want, got := common.HexToHash(checksums[i]), common.Hash(sha256.Sum256(f)); want != got {
 			t.Fatalf("checksum %d does not match: got %s, want %s", i, got, want)
 		}
-		r, err := era.NewReader(bytes.NewReader(f))
+
+		r, err := era.NewReader(&loggingSeeker{bytes.NewReader(f)})
 		if err != nil {
 			t.Fatalf("error making era reader: %v", err)
 		}
 		for j := 0; ; j += 1 {
+			fmt.Printf("Now reading block + receipts for block %d\n", j)
 			block, receipts, err := r.Read()
+			fmt.Printf("Finished reading block %d\n", j)
+
 			if err == io.EOF {
 				break
 			} else if err != nil {
@@ -169,3 +174,17 @@ func TestHistoryImportAndExport(t *testing.T) {
 		t.Fatalf("imported chain does not match expected, have (%d, %s) want (%d, %s)", have.Number, have.Hash(), want.Number, want.Hash())
 	}
 }
+
+type loggingSeeker struct {
+	r era.ReadAtSeeker
+}
+
+func (l *loggingSeeker) ReadAt(p []byte, off int64) (n int, err error) {
+	fmt.Printf("ReadAt( %d : %d)\n", off, off+int64(len(p)))
+	return l.r.ReadAt(p, off)
+
+}
+func (l *loggingSeeker) Seek(offset int64, whence int) (int64, error) {
+	fmt.Printf("Seek(offset %d, whence %d)\n", offset, whence)
+	return l.r.Seek(offset, whence)
+}
```
And then run the `HistoryTest`. You will see, e.g. 
```
Now reading block + receipts for block 12
ReadAt( 7755 : 7763)
ReadAt( 5608 : 5616)
ReadAt( 5616 : 5868)
ReadAt( 7755 : 7763)
ReadAt( 5608 : 5616)
ReadAt( 5868 : 5876)
ReadAt( 5876 : 6006)
ReadAt( 7755 : 7763)
ReadAt( 5608 : 5616)
ReadAt( 5868 : 5876)
ReadAt( 6006 : 6014)
ReadAt( 6014 : 6066)
Finished reading block 12
```
So the `r.Read()` causes ~ 12 syscalls. It reeds some portions, like `7755 : 7763`, `5608 : 5616`  three times, and other sections two times. And even if the filesystem read is cached, I'm pretty sure we also spin up a snappyreader anew to decode it. 

So there's definitely some optimizations up for grabs there.

holiman commented on internal/era/era.go:
```diff
@@ -0,0 +1,336 @@
+}
+
+// ReadBlock reads the block number n.
+func (r *Reader) ReadBlock(n uint64) (*types.Block, error) {
```
One thing that this implementation does quite a bit, is handle full arrays. This is often the simplest way to deal with data, but leads to large overhead in processing. 

Like, a file is read in full from disk. It is marshalled and copied into a slice in a struct. Then that data is decompressed into a new buffer. Then that buffer given to `rlp` to marshal into a `block` struct. 

A lot of memory copying/allocation can be avoided if e.g. `io.Reader`s are used in the pipeline. Here's one example where an `io.Reader` could be used instead: 

```golang
func (r *Reader) At(n uint64, skip int) (io.Reader, error) {
	e, err := r.readEntry(n, skip)
	if err != nil {
		return nil, err
	}
	return snappy.NewReader(bytes.NewReader(e.Value)), nil
}

// ReadBlock reads the block number n.
func (r *Reader) ReadBlock(n uint64) (*types.Block, error) {
	header, err := r.ReadHeader(n)
	if err != nil {
		return nil, err
	}
	var body types.Body
	input, err := r.At(n, 1)
	if err != nil {
		return nil, err
	}
	if rlp.Decode(input, &body); err != nil {
		return nil, err
	}
	return types.NewBlockWithHeader(header).WithBody(body.Transactions, body.Uncles), nil
}
```
The `snappy.Reader` is passed directly to `rlp.Decode`, avoiding the `io.ReadAll` and managing the buffer explicitly. 

I didn't commit this code, because it might conflict with other potential changes that you might want to do in order to shave down on the number of reads (as I commented on yesterday). Just posting this as a potential direction to take things into.

holiman replied:
something like this `e2store/e2store.go` maybe ? 
```golang
// ReadAt reads one Entry from r at the specified offset.
func (r *Reader) ReaderAt(expectedType uint16, off int64) (io.Reader, error) {
	typ, length, err := r.ReadMetadataAt(off)
	if err != nil {
		return nil, err
	}
	if typ != expectedType {
		return nil, fmt.Errorf("wrong type, want %d have %d", expectedType, typ)
	}
	if length > valueSizeLimit {
		return nil, fmt.Errorf("item larger than item size limit %d: have %d", valueSizeLimit, length)
	}
	return io.NewSectionReader(r.r, off+headerSize, int64(length)), nil
}
```

holiman replied:
So by using a pipeline like this, the rlp marshaller would 'pull' from the snappy decoder, which would 'pull' from the `e2store` reader, which would 'pull' from the file. And you'd have a pull-based input flow, instead of passing around slices of bytes.

lightclient replied:
Really good insight - thanks for this. I've added the recommend function in c1a1f93a4c900318df6da88060703a5d67af2cbb.

lightclient commented:
Okay just updated the interface to be more an iterator-style, and along the way optimized the read pattern when accessing several records from a single block entry. This means the reads for block 12 look like this:

```
Now reading block + receipts for block 12
ReadAt( 7755 : 7763 )
ReadAt( 5608 : 5616 )
ReadAt( 5616 : 5868 )
ReadAt( 5868 : 5876 )
ReadAt( 5876 : 6006 )
ReadAt( 6006 : 6014 )
ReadAt( 6014 : 6066 )
ReadAt( 6066 : 6074 )
ReadAt( 6074 : 6106 )
done reading block
```

Duplicates are now removed and so syscalls are down from ~12 to ~9.

holiman commented on cmd/era/main.go:
```diff
@@ -0,0 +1,325 @@
+	//   * the accumulator is correct by recomputing it locally,
+	//     which verifies the blocks are all correct (via hash)
+	//   * the receipts root matches the value in the block
+	for j := 0; it.Next(); j++ {
```
```suggestion
	for it.Next() {
```
Wouldn't that be neater -- dropping the `j`? And then you'd have to add some sort of accessor for the iterator Key, which in this case probably is a `Number()`, to use instead of the `j` ?

lightclient replied:
Yeah that is a good idea! Now with `Number()` it isn't necessary.

holiman commented on internal/era/e2store/e2store.go:
```diff
@@ -0,0 +1,196 @@
+// record store the type (2 bytes), the length (4 bytes), and some reserved
+// data (2 bytes). The remaining bytes store b.
+func (w *Writer) Write(typ uint16, b []byte) (int, error) {
+	buf := make([]byte, headerSize+len(b))
+	binary.LittleEndian.PutUint16(buf, typ)
+	binary.LittleEndian.PutUint32(buf[2:], uint32(len(b)))
+	copy(buf[8:], b)
+	return w.w.Write(buf)
```
Is `b` here expected to be fairly large? If so, this is a bit suboptimal: you temporarily double the mem consumption, since you allocate `buf` and copy into, before you write into `w`. A more mem-friendly way to do it is 

```suggestion
	buf := make([]byte, headerSize)
	binary.LittleEndian.PutUint16(buf, typ)
	binary.LittleEndian.PutUint32(buf[2:], uint32(len(b)))
	if n, err := w.w.Write(buf); err != nil{
		return n, err
	} 
	return w.w.Write(b)
```

lightclient replied:
Ah right that's true. The values are as large as our block bodies, so this is a lot to be copying. I've updated so that it doesn't need to temporarily have that double allocation.

lightclient commented:
I'm still planning to implement the lazy reading btw! Then I think I will have addressed all of your review feedback 🙏

lightclient commented:
Okay `e2store` now has a `ReaderAt` function which returns an `io.SectionReader` for that entry's value!

holiman commented on internal/era/iterator.go:
```diff
@@ -0,0 +1,163 @@
+	}
+	var n int64
+	if it.Header, n, it.err = newSnappyReader(it.e.s, TypeCompressedHeader, off); it.err != nil {
+		return true
```
Isn't it is possible that `it.Header` becomes set to something (`nil`?), and `it.Body` is left set as something else (the previous block)? 
If so `return types.NewBlockWithHeader(&header).WithBody(body.Transactions, body.Uncles), nil` in `Block()` --  will return something wonky

lightclient replied:
Yeah was having a hard time thinking exactly how the interface should be. The `bool` return value should denote whether `Next()` can be called again. As long as the block index is intact and the caller desires, I think they should be able to continue calling `Next()` even if a section reader for the header, body, etc can't be returned. But in general, I don't know why they would. I think anytime the iterator has an error, the caller will most likely want to abort as the whole era is corrupted.
--
It's definitely possible that for one of the values to get set to `nil` and the others to not be updated. Generally I think the caller should check `it.Error()` before getting a value from the iterator, but I guess it doesn't hurt to protect them some more by `nil`ing them out in advent of an error.

holiman commented on internal/era/iterator.go:
```diff
@@ -0,0 +1,163 @@
+		header types.Header
+		body   types.Body
+	)
+	if err := rlp.Decode(it.inner.Header, &header); err != nil {
```
shouldn't you check for `it.inner.err` at some point?

lightclient replied:
I think it should be up to the caller to check `it.inner.err`, but maybe what I could do here is check if `it.inner.Header` or `it.inner.Body` are `nil` and return an error to that effect?

holiman commented on internal/era/iterator.go:
```diff
@@ -0,0 +1,163 @@
+}
+
+// Receipts returns the receipts for the iterator's current position.
+func (it *Iterator) Receipts() (types.Receipts, error) {
```
check `it.inner.err`?

lightclient commented on internal/era/iterator.go:
```diff
@@ -0,0 +1,163 @@
+	inner *RawIterator
+}
+
+// NewIterator returns a new Iterator instance.
```
Good point, I will update it. You must call `Next` after getting a new iterator. I'll make a note of the other cases too.

fjl commented:
Suggested merging date: Feb 6 2024

lightclient commented:
latest runs:
export 5h50m
import 9h58m

so-not-lightclient commented:
I'm so excited for this feature ! I really hope it is done marinating soon.

holiman commented on internal/era/iterator.go:
```diff
@@ -0,0 +1,197 @@
+// Receipts, TotalDifficulty will be set to nil in the case returning false or
+// finding an error and should therefore no longer be read from.
+func (it *RawIterator) Next() bool {
+	// Clear old errors.
+	it.err = nil
```
... Interesting... Why though?

lightclient replied:
I guess it could be part of `clear()`, but it right now if `err` isn't clear out, some paths return false without setting error.

MariusVanDerWijden commented on cmd/era/main.go:
```diff
@@ -0,0 +1,324 @@
+		if rr != block.ReceiptHash() {
+			return fmt.Errorf("receipt root in block %d mismatch: want %s, got %s", block.NumberU64(), block.ReceiptHash(), rr)
+		}
+		hashes = append(hashes, block.Hash())
```
Do you want to verify the withdrawals root in the future as well?

lightclient replied:
The `era1` format is not intended to ever support withdrawals. We'll want to use the standard `era` format currently exported by nimbus.

ryanschneider replied:
so if I want to restore an archive node using these files in a post-4444 world, would I use .era1 files up to the merge and then .era files for merge up some post-weak subjectivity period checkpoint?  Would this all be done on the EL, or would I import that .era files in my CL?

lightclient replied:
It hasn't really been determined how the post-merge `era` files will be ingested. Two obvious paths:

1. CL imports `era` files and pushes payloads directly to EL.
2. CL imports beacon chain data from `era` files and EL imports execution chain data from same `era` files.

Obvious 1) seems preferred from a UX standpoint, but not sure it will be performant enough. Plus, some clients may opt to use `era` in-place as an ancient store. In which case you could just point both clients to the directory containing the `era` files.

MariusVanDerWijden commented on cmd/geth/chaincmd.go:
```diff
@@ -354,7 +383,97 @@ func exportChain(ctx *cli.Context) error {
+			}
+		}
+		if len(networks) == 0 {
+			return fmt.Errorf("no era1 files found in %s", dir)
```
This will also print "no era1 files found" if they are under a different name. In my case I tried to run this pr on devnet-9 whcih will produce files like this: `unknown-00001-40745657.era1`. I can't import them anymore

MariusVanDerWijden replied:
maybe we can add something s.th. if no other networks are found, it will try to import the blocks using the "unknown" network

ryanschneider replied:
what about using the chain id as the prefix instead of the network name?  Or maybe something like `<vanity-network-name>-<chain-id>` where the vanity name _should_ match the chain id but it just warns if it doesn't?

lightclient replied:
@MariusVanDerWijden: it is a fair argument that era1 should probably have a bit better support for unknown networks. However, we don't really have new unmerged networks these days. This makes me worried that people will attempt to do as you have done with devnet-9, export it as era1. We should really push people in non-pow networks to use regular era format as it has all the data relevant to the chain (e.g. the beacon data).

I looked a bit into more generic support for era1 to do what you were trying to do, but my feeling is the gate is actually kinda of useful. Let me know if you still disagree.

@ryanschneider: what will the chain id do for us in this case? Just a more robust identifier? As stated above, I don't think we need to generically support networks because the number of PoW networks we care about is dwindling (in a few years, only mainnet will be left).

MariusVanDerWijden commented:
I think I found another small issue, I exported the chain twice, once with `export-history out 0 15000` and once with `export-history out 0 42000` thus I have the following files: `unknown-00000-6323cbe0.era1  unknown-00001-40745657.era1  unknown-00001-8ffaa878.era1  unknown-00002-4c296dc4.era1` (note the duplicated epoch 1). If I try to import it, I get the following error:
`error reading out: missing epoch 2`. I guess it is because of the duplicated epoch 1

MariusVanDerWijden commented on internal/era/accumulator.go:
```diff
@@ -0,0 +1,90 @@
+}
+
+// reverseOrder reverses the byte order of a slice.
+func reverseOrder(b []byte) []byte {
```
this assumes that the byte slice is exactly 32 bytes long, right?
I think its fine bigToBytes32, but it is also used for the total difficulty (which is extended to 32 bytes in the writer) but adding at a comment pointing to this assumption would be good imo

MariusVanDerWijden commented:
Ci is red

njgheorghita commented on internal/era/e2store/e2store_test.go:
```diff
@@ -0,0 +1,150 @@
+		{ // basic valid decoding
+			have: "ffff000000000000",
+		},
+		{ // basic valid decoding
```
`// basic invalid decoding`

MariusVanDerWijden commented on cmd/era/main.go:
```diff
@@ -0,0 +1,324 @@
+	//      the blocks are all correct (via hash)
+	//
+	// The attributes 1), 2), and 3) are checked for each block. 4) and 5) require
+	// accumulation accross the entire set and are verified at the end.
```
```suggestion
	// accumulation across the entire set and are verified at the end.
```

kdeme reviewed (commented):
I'm in the process of implementing era1 for Nimbus and while doing so it came to my attention that Snappy compression is non-deterministic. Hence, I don't think we can use the checksums verification cross-client/implementation.
So I think that part should be removed from the specification when it gets more formalized.

Some smaller suggestions to the specification in the PR description:

- It should be mentioned that TotalDifficulty's uint256 must be LE encoded (as it is not rlp or ssz encoded).
- Change the name of Accumulator to AccumulatorRoot to better represent the actual data.



And lastly I was wondering if it would make sense to add an Index also for the accumulator root? Similar as is done in consensus era files for the full state.

Unless I misunderstand the specs, I think currently one would have to jump all the records until its hits the Accumulator root. (Well, one could read the BlockIndex and jump immediately to the 8191th block tuple records and skip those, but that feels a bit like a hack).

Regarding use cases of getting the Accumulator root individually. Typically it would be read after also iterating over all the block-tuples to actually build the Accumulator for verification. So in that scenario it is not really needed.

But one use case could be when you want to verify the accumulator root with values you have out-of-band, before you get into reading/verifying all the data.

lightclient commented:
>  Hence, I don't think we can use the checksums verification cross-client/implementation.
So I think that part should be removed from the specification when it gets more formalized.

Agree that they cannot be used for cross-client verification. I moved around the note about them and added they are not part of the spec. I do intend to share them alongside our era1 files though for convenience of verification. I think other clients can do the same, or we can decide on a single implementation to share. This works for era1 because we only need to do it once, we aren't constantly generating them each time.

> It should be mentioned that TotalDifficulty's uint256 must be LE encoded (as it is not rlp or ssz encoded).

Updated to mention it is LE encoded.

> Change the name of Accumulator to AccumulatorRoot to better represent the actual data.

Updated.

> And lastly I was wondering if it would make sense to add an Index also for the accumulator root? Similar as is done in consensus era files for the full state.

I think we should keep things simple. As you said, you can technically just traverse to before the block index if you really need the performance.

arnetheduck commented:
> or we can decide on a single implementation to share

...and version. 

I don't quite see the use case of "simple verification" here, in the sense that anyone ingesting the files will have the means to decode them as well _and_ the data is already covered by crc32 for "trivial" verification (because snappy framed).

> This works for era1 because we only need to do it once, we aren't constantly generating them each time.

An independent implementation _should_ be able to generate a valid era1 file to verify that the data was correctly generated by the other implementation - this implies that the spec _should_ have a section on tests that the files should pass (similar to era for eth2)

kdeme commented:
> I think we should keep things simple. As you said, you can technically just traverse to before the block index if you really need the performance.

Fair enough, I wasn't sure about this suggestion myself. I've now implemented it by just reversing the size of the record from the block index position.

lightclient commented:
> I don't quite see the use case of "simple verification" here, in the sense that anyone ingesting the files will have the means to decode them as well and the data is already covered by crc32 for "trivial" verification (because snappy framed).

It takes a very long time to verify the data integrity because you need to recompute the tx hash and receipt hash. On the order of hours. I don't have a recent benchmark. So the checksum is extremely fast verification that data is correct.

This is especially useful if you're going to just use era1 as a drop in db backend for block look ups. No need to spend hours validating / ingesting.

> An independent implementation should be able to generate a valid era1 file to verify that the data was correctly generated by the other implementation - this implies that the spec should have a section on tests that the files should pass (similar to era for eth2)

Not sure exactly which tests you're referring to? I can add something about clients should be able to i) compute the same accumulator root for the era1 and iii) verify each total difficulty is correct.

arnetheduck commented:
> It takes a very long time to verify the data integrity because you need to recompute the tx hash and receipt hash. 

why does this take significantly longer than computing a file hash? ie it's a similar order of magnitude of data to be hashed.

> This is especially useful if you're going to just use era1 as a drop in db backend for block look ups. 

We ended up doing this for era but without any such intermediate verification - basically even just reading a large archive like this takes time and because the consumer must do a verification anyway, typically (say if you're serving the data over the wire) it doesn't greatly matter for the provider. 

If on the other hand _you_ want to use the data (ingest it), it's not a bad idea to check that it is well-formed and leads to the actual block you expect.

The in-between seems like it'll still take a not insignificant amount of time, is non-deterministic and says more or less that the file transfer worked with a bit higher confidence than the already-present crc32.

FWIW, I'm curious about this case for the simple reason that we _could_ add a header in the file that contains a hash as a new "type", but ...

What we do for verification in consensus land is read the header of the beacon state which contains the block tree hashes then verify those - if I were to guess, this takes 2-3x longer than simply hashing the data, which in the grand scheme of things isn't much of a difference.

The other trick, taken from the crc32, would be to focus on the uncompressed data and hash that - this would be compatible across all implementations (obviously).

lightclient commented:
> why does this take significantly longer than computing a file hash? ie it's a similar order of magnitude of data to be hashed.

There is a lot of copying, reflection, and serializing to compute the roots. We have to read the block body into our block body structure, then get the transactions out of it, re-serialize, compute root, etc. I would be curious how fast other clients can do this. Of course it can be done faster, but at a rather large cost to code complexity.

--

I don't really understand what you mean in your second comment by "intermediate verification" and "in-between". Do you think there needs to be an additional type in the era1 format?

holiman reviewed (approved):
Let's goo

arnetheduck commented:
> I don't really understand what you mean in your second comment by "intermediate verification" and "in-between". Do you think there needs to be an additional type in the era1 format?

I mean that verifying a sha256sum of a file generated in a non-deterministic way (due to snappy) provides little value in the end - it still takes time, doesn't bring cross-client guarantees or "data idempotency" meaning we can't post those hashes as "universal truth" without trusting a single-client implementation and for the end user, it brings limited value in the sense that transmission errors will get detected anyway and for "real" verification of the file, a more elaborate process is needed regardless ("this data corresponds to the consensus of the merge block").

This is probably not a hill to die on, but tying the hashes to a particular go-snappy version _and_ its particular usage in geth feels off.

lightclient commented:
I'm not proposing we make the checksums a universal source of truth. That is what the accumulator root is for. It's just a convenient option for go-ethereum users and doesn't impact the standard. Other clients do not need to do this.

kdeme commented on internal/era/builder.go:
```diff
@@ -0,0 +1,228 @@
+		index = make([]byte, 16+count*8)
+	)
+	binary.LittleEndian.PutUint64(index, *b.startNum)
+	// Each offset is relative from the position it is encoded in the
```
This is a difference with the spec/implementation of consensus layer era files: `SlotIndex records store offsets, in bytes, from the beginning of the index record to the beginning of the corresponding data at that slot.`

I assume this is intentional (I think this is what the sentence from the spec in the PR means: `Every index is a defined relative to index's location in the file.`, but it isn't all that clear to me.) 

I'm not stating here that one or the other is better, but it is slightly annoying that they are different imo.

lightclient replied:
This is an oversight on my part. I think they should aligned, I just misunderstood that portion of the spec as relative to the index being written. I will fix.

kdeme replied:
Ok, great!

Yes, the offsets are basically always relative to the same `BlockIndex` record starting position.

You can see the difference between the two in this comment/diff: https://github.com/status-im/nimbus-eth1/pull/1998#discussion_r1479810389

lightclient replied:
Will do this in a separate PR.

lightclient commented:
![image](https://github.com/ethereum/go-ethereum/assets/14004106/e5a52660-fe6a-4ffb-b063-f0fa553291de)
